File: src/types/charDetail.ts
// 角色详情数据类型定义

export interface CharDetailResponse {
  code: number;
  message: string;
  timestamp: string;
  data: CharDetailDataWrapper;
}

export interface CharDetailDataWrapper {
  detail: CharDetailData;
}

export interface CharDetailData {
  base: BaseInfo;
  chars: CharacterItem[];
  dungeon?: any;
  bpSystem?: BpSystem;
  dailyMission?: DailyMission;
  weeklyMission?: WeeklyMission;
  spaceShip?: any;
  domain?: any;
  quickaccess?: any;
  config?: any;
  achieve?: AchieveData;
  currentTs?: number;
}

export interface BaseInfo {
  serverName: string;
  roleId: string;
  name: string;
  createTime: string;
  saveTime: string;
  lastLoginTime: string;
  exp: number;
  level: number;
  worldLevel: number;
  gender: number;
  avatarUrl: string;
  mainMission: MainMission;
  charNum: number;
  weaponNum: number;
  docNum: number;
}

export interface MainMission {
  id: string;
  description: string;
}

export interface CharacterItem {
  charData: CharacterData;
  id: string;
  level?: number;
  evolvePhase?: number;
  potentialLevel?: number;
  userSkills?: any;
  bodyEquip?: any;
  armEquip?: any;
  firstAccessory?: any;
  secondAccessory?: any;
  weapon?: any;
  tacticalItem?: any;
  talent?: TalentNodes;
  wikiItemId?: string;
}

export interface CharacterData {
  id: string;
  name: string;
  avatarSqUrl: string;
  avatarRtUrl: string;
  rarity: RarityInfo;
  profession: ProfessionInfo;
  property: PropertyInfo;
  weaponType: WeaponTypeInfo;
  skills: SkillInfo[];
  illustrationUrl: string;
  tags: string[];
  abilityTalents: TalentInfo[];
  combatTalents: TalentInfo[];
  cultivationTalents?: TalentInfo[];
}

export interface RarityInfo {
  key: string;
  value: string; // "6", "5", etc.
}

export interface ProfessionInfo {
  key: string;
  value: string; // "术师", "近卫", etc.
}

export interface PropertyInfo {
  key: string;
  value: string; // "寒冷", "物理", etc.
}

export interface WeaponTypeInfo {
  key: string;
  value: string; // "手铳", "剑", etc.
}

export interface SkillInfo {
  id: string;
  name: string;
  type: SkillTypeInfo;
  property: PropertyInfo;
  iconUrl: string;
  desc: string;
  descParams: Record<string, any>;
  descLevelParams: Record<string, any>;
}

export interface SkillTypeInfo {
  key: string;
  value: string; // "普通攻击", "战技", etc.
}

export interface TalentInfo {
  id: string;
  name: string;
  iconUrl: string;
  desc: string;
  descParams?: Record<string, any>;
  lockedIconUrl: string;
}

export interface TalentNodes {
  latestBreakNode: string;
  attrNodes: string[]; // 能力天赋节点 ID
  latestPassiveSkillNodes: string[]; // 战斗天赋节点 ID
  latestFactorySkillNodes: string[]; // 制造天赋节点 ID
  latestSpaceshipSkillNodes: string[]; // 培养天赋节点 ID
}

// ====== Wiki 相关类型 ======

/** Wiki 分类目录 */
export interface WikiCatalog {
  code: number;
  message: string;
  timestamp: string;
  data: WikiCatalogData;
}

export interface WikiCatalogData {
  catalog: WikiCatalogEntry[];
}

export interface WikiCatalogEntry {
  typeMain: number;
  typeSub: WikiTypeSub[];
}

export interface WikiTypeSub {
  typeSubId: number;
  typeSubName: string;
  items: WikiCatalogItem[];
}

/** Wiki 目录中的单个物品条目 */
export interface WikiCatalogItem {
  itemId: string;
  name: string;
  iconUrl?: string;
  rarity?: number;
  type?: string;
  [key: string]: any;
}

/** Wiki 物品详情响应 */
export interface WikiItemResponse {
  code: number;
  message: string;
  timestamp: string;
  data: WikiItemData;
}

export interface WikiItemData {
  item: WikiItemDetail;
}

/** 文档内联元素 */
export interface WikiInlineElement {
  kind: "text" | "link" | "image";
  text?: { text: string; href?: string };
  image?: { src: string; alt?: string };
}

/** 文档块 */
export interface WikiDocumentBlock {
  id: string;
  parentId: string;
  align?: string;
  kind: "text" | "heading3" | "horizontalLine" | "image" | "quote" | "table";
  text?: {
    inlineElements: WikiInlineElement[];
    kind: "body";
  };
  image?: { src: string; alt?: string };
  type?: "common";
  tableList?: any[];
  tabList?: WikiDocTab[];
  tabDataMap?: Record<string, WikiTabData>;
}

/** 文档标签页定义 */
export interface WikiDocTab {
  tabId: string;
  title: string;
  icon: string;
}

/** 文档标签页数据 */
export interface WikiTabData {
  intro: WikiTabIntro | null;
  content: string;
  audioList: any[];
}

/** 标签页介绍头 */
export interface WikiTabIntro {
  name: string;
  type: string;
  imgUrl?: string;
  description?: string;
}

/** 文档条目（block 型） */
export interface WikiBlockDocumentEntry {
  id: string;
  blockIds: string[];
  blockMap: Record<string, WikiDocumentBlock>;
}

/** 文档条目（tab 型） */
export interface WikiTabDocumentEntry {
  type: "common";
  tableList: any[];
  tabList: WikiDocTab[];
  tabDataMap: Record<string, WikiTabData>;
}

/** 文档条目 */
export type WikiDocumentEntry = WikiBlockDocumentEntry | WikiTabDocumentEntry;

/** 关联对象 */
export interface WikiAssociate {
  id: string;
  name: string;
  type: string;
  dotType: string;
}

/** 子类型映射 */
export interface WikiSubTypeEntry {
  subTypeId: string;
  value: string;
}

/** 主/子分类 */
export interface WikiTypeInfo {
  id: string;
  name: string;
  status?: number;
  position?: number;
  typeSub?: any[];
  fatherTypeId?: string;
  style?: number;
  icon?: string;
  items?: any[];
  filterTagTree?: any[];
}

/** Wiki 物品详情 */
export interface WikiItemDetail {
  itemId: string;
  name: string;
  brief?: string;
  lang?: string;
  document: {
    documentMap: Record<string, WikiDocumentEntry>;
    associate: WikiAssociate;
    dotType: string;
    subTypeList: WikiSubTypeEntry[];
    composite: any;
    disableCoverShowInDetail: boolean;
  };
  mainType: WikiTypeInfo;
  subType: WikiTypeInfo;
  status: number;
  publishedAtTs: string;
  lastAuditPassedAt: string;
  tagIds: string[];
  createdUser?: any;
  lastUpdatedUser?: any;
}

export interface BpSystem {
  curLevel: number;
  maxLevel: number;
}

export interface DailyMission {
  dailyActivation: number;
  maxDailyActivation: number;
}

export interface WeeklyMission {
  score: number;
  total: number;
}

// ====== 成就/奖章相关类型 ======

export interface AchievementData {
  id: string;
  name: string;
  initIcon: string;
  reforge2Icon: string;
  reforge3Icon: string;
  platedIcon: string;
  cateName: string;
  canCertify: boolean;
  cate: string;
  initLevel: number;
}

export interface AchieveMedal {
  achievementData: AchievementData;
  level: number;
  isPlated: boolean;
  obtainTs: string;
}

export interface AchieveData {
  achieveMedals: AchieveMedal[];
  display?: Record<string, string>;
}

File: src-tauri/src/models/char_detail.rs
use serde::{Deserialize, Serialize};

/// 顶层响应包装
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct CharDetailResponse {
    pub code: i32,
    pub message: String,
    pub timestamp: String,
    pub data: CharDetailDataWrapper,
}

/// 数据包装器
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct CharDetailDataWrapper {
    pub detail: CharDetailData,
}

/// 主数据结构 - 包含所有角色详情信息
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct CharDetailData {
    pub base: BaseInfo,
    #[serde(default, deserialize_with = "deserialize_chars_fallback")]
    pub chars: Vec<CharacterItem>,
    #[serde(default)]
    pub dungeon: Option<serde_json::Value>,
    #[serde(default)]
    pub bp_system: Option<BpSystem>,
    #[serde(default)]
    pub daily_mission: Option<DailyMission>,
    #[serde(default)]
    pub weekly_mission: Option<WeeklyMission>,
    #[serde(default)]
    pub space_ship: Option<serde_json::Value>,
    #[serde(default)]
    pub domain: Option<serde_json::Value>,
    #[serde(default)]
    pub quickaccess: Option<serde_json::Value>,
    #[serde(default)]
    pub config: Option<serde_json::Value>,
    #[serde(default)]
    pub achieve: Option<serde_json::Value>,
    #[serde(default, deserialize_with = "deserialize_string_to_i64")]
    pub current_ts: Option<i64>,
}

// 自定义反序列化器：将字符串或数字转换为 i64
fn deserialize_string_to_i64<'de, D>(deserializer: D) -> Result<Option<i64>, D::Error>
where
    D: serde::Deserializer<'de>,
{
    use serde::de::Error;

    // 尝试解析为 serde_json::Value
    let value: Option<serde_json::Value> = Option::deserialize(deserializer)?;

    match value {
        None => Ok(None),
        Some(serde_json::Value::Number(n)) => Ok(n.as_i64()),
        Some(serde_json::Value::String(s)) => s
            .parse::<i64>()
            .map(Some)
            .map_err(|e| Error::custom(format!("Failed to parse string '{}' as i64: {}", s, e))),
        _ => Err(Error::custom("Expected a number or string")),
    }
}

// 自定义反序列化器：如果 chars 解析失败，返回空数组
fn deserialize_chars_fallback<'de, D>(deserializer: D) -> Result<Vec<CharacterItem>, D::Error>
where
    D: serde::Deserializer<'de>,
{
    use serde::de::Error;

    // 尝试解析为 Vec<CharacterItem>
    let result: Result<Vec<CharacterItem>, _> = Vec::deserialize(deserializer);

    match result {
        Ok(chars) => Ok(chars),
        Err(e) => {
            // 记录错误但返回空数组，避免整个解析失败
            eprintln!(
                "Warning: Failed to deserialize chars array: {}. Returning empty array.",
                e
            );
            Ok(Vec::new())
        }
    }
}

/// 基础玩家信息
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct BaseInfo {
    pub server_name: String,
    pub role_id: String,
    pub name: String,
    pub create_time: String,
    pub save_time: String,
    pub last_login_time: String,
    pub exp: i64,
    pub level: i32,
    pub world_level: i32,
    pub gender: i32,
    pub avatar_url: String,
    pub main_mission: MainMission,
    pub char_num: i32,
    pub weapon_num: i32,
    pub doc_num: i32,
}

/// 主线任务信息
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct MainMission {
    pub id: String,
    pub description: String,
}

/// 干员项（API 返回的格式）
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct CharacterItem {
    #[serde(default)]
    pub char_data: Option<CharacterData>,
    #[serde(default)]
    pub id: Option<String>,
    #[serde(default)]
    pub level: Option<i32>,
    #[serde(default)]
    pub evolve_phase: Option<i32>,
    #[serde(default)]
    pub potential_level: Option<i32>,
    #[serde(default)]
    pub user_skills: Option<serde_json::Value>,
    #[serde(default)]
    pub body_equip: Option<serde_json::Value>,
    #[serde(default)]
    pub wiki_item_id: Option<String>,
    #[serde(default)]
    pub talent: Option<TalentNodes>,
}

/// 天赋节点信息
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct TalentNodes {
    pub latest_break_node: String,
    pub attr_nodes: Vec<String>,                   // 能力天赋节点 ID
    pub latest_passive_skill_nodes: Vec<String>,   // 战斗天赋节点 ID
    pub latest_factory_skill_nodes: Vec<String>,   // 制造天赋节点 ID
    pub latest_spaceship_skill_nodes: Vec<String>, // 培养天赋节点 ID
}

/// 干员详细数据
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct CharacterData {
    #[serde(default)]
    pub id: Option<String>,
    #[serde(default)]
    pub name: Option<String>,
    #[serde(default)]
    pub avatar_sq_url: Option<String>,
    #[serde(default)]
    pub avatar_rt_url: Option<String>,
    #[serde(default)]
    pub rarity: Option<RarityInfo>,
    #[serde(default)]
    pub profession: Option<ProfessionInfo>,
    #[serde(default)]
    pub property: Option<PropertyInfo>,
    #[serde(default)]
    pub weapon_type: Option<WeaponTypeInfo>,
    #[serde(default)]
    pub skills: Option<Vec<SkillInfo>>,
    #[serde(default)]
    pub illustration_url: Option<String>,
    #[serde(default)]
    pub tags: Option<Vec<String>>,
    #[serde(default)]
    pub ability_talents: Option<Vec<TalentInfo>>,
    #[serde(default)]
    pub combat_talents: Option<Vec<TalentInfo>>,
    #[serde(default)]
    pub cultivation_talents: Option<Vec<TalentInfo>>,
}

/// 稀有度信息
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct RarityInfo {
    pub key: String,
    pub value: String,
}

/// 职业信息
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct ProfessionInfo {
    pub key: String,
    pub value: String,
}

/// 属性信息
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct PropertyInfo {
    pub key: String,
    pub value: String,
}

/// 武器类型信息
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct WeaponTypeInfo {
    pub key: String,
    pub value: String,
}

/// 技能信息
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct SkillInfo {
    pub id: String,
    pub name: String,
    #[serde(rename = "type")]
    pub skill_type: SkillTypeInfo,
    pub property: PropertyInfo,
    pub icon_url: String,
    pub desc: String,
    pub desc_params: serde_json::Value,
    pub desc_level_params: serde_json::Value,
}

/// 技能类型信息
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct SkillTypeInfo {
    pub key: String,
    pub value: String,
}

/// 天赋信息
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct TalentInfo {
    pub id: String,
    pub name: String,
    pub icon_url: String,
    pub desc: String,
    pub desc_params: Option<serde_json::Value>,
    pub locked_icon_url: String,
}

/// 战斗通行证系统
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct BpSystem {
    pub cur_level: i32,
    pub max_level: i32,
}

/// 每日任务
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct DailyMission {
    pub daily_activation: i32,
    pub max_daily_activation: i32,
}

/// 每周任务
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct WeeklyMission {
    pub score: i32,
    pub total: i32,
}
